File: src/lib/domain/value-objects/subject-category.ts
import type { FieldCategory } from "./field-category.ts";

export type SubjectCategoryKind =
	// R2-R5 (AY2020-2023, Reiwa 2-5) categories
	| "common-education/primary"
	| "common-education/liberal/field"
	| "common-education/liberal/foreign-language"
	| "common-education/liberal/career"
	// R6+ (AY2024 onward, Reiwa 6+) introductory course groups
	| "common-education/introductory/core-learning"
	| "common-education/introductory/core-english"
	| "common-education/introductory/foreign-language"
	| "common-education/introductory/math-ai"
	// R6+ liberal course group: 7 fields
	| "common-education/liberal-group/life"
	| "common-education/liberal-group/health-sports"
	| "common-education/liberal-group/career"
	| "common-education/liberal-group/arts"
	| "common-education/liberal-group/humanities-social"
	| "common-education/liberal-group/natural-science"
	| "common-education/liberal-group/complex"
	// Seminar (both regimes)
	| "seminar/1-2"
	| "seminar/3-4/spring"
	| "seminar/3-4/fall"
	| "seminar/5-6-thesis"
	// R2-R5 platform structure
	| "platform/basic-a"
	| "platform/basic-b"
	| "platform/foreign-language"
	| "platform/advanced"
	// R6+ platform structure
	| "platform/faculty-common"
	| "platform/humanities"
	| "platform/global-studies"
	| "platform/social-science"
	// Elective (both regimes)
	| "elective/own-course"
	| "elective/other-course"
	| "elective/other-faculty"
	| "unknown";

export type SubjectCategory =
	| { readonly kind: "common-education/primary" }
	| {
			readonly kind: "common-education/liberal/field";
			readonly field: FieldCategory;
	  }
	| {
			readonly kind: "common-education/liberal/foreign-language";
			readonly language: string;
	  }
	| { readonly kind: "common-education/liberal/career" }
	// R6+ introductory course groups
	| { readonly kind: "common-education/introductory/core-learning" }
	| { readonly kind: "common-education/introductory/core-english" }
	| {
			readonly kind: "common-education/introductory/foreign-language";
			readonly language: string;
	  }
	| { readonly kind: "common-education/introductory/math-ai" }
	// R6+ liberal course group: 7 fields
	| { readonly kind: "common-education/liberal-group/life" }
	| { readonly kind: "common-education/liberal-group/health-sports" }
	| { readonly kind: "common-education/liberal-group/career" }
	| { readonly kind: "common-education/liberal-group/arts" }
	| { readonly kind: "common-education/liberal-group/humanities-social" }
	| { readonly kind: "common-education/liberal-group/natural-science" }
	| { readonly kind: "common-education/liberal-group/complex" }
	| { readonly kind: "seminar/1-2" }
	| { readonly kind: "seminar/3-4/spring" }
	| { readonly kind: "seminar/3-4/fall" }
	| { readonly kind: "seminar/5-6-thesis" }
	| { readonly kind: "platform/basic-a" }
	| { readonly kind: "platform/basic-b" }
	| { readonly kind: "platform/foreign-language" }
	| { readonly kind: "platform/advanced" }
	// R6+ platform
	| { readonly kind: "platform/faculty-common" }
	| { readonly kind: "platform/humanities" }
	| { readonly kind: "platform/global-studies" }
	| { readonly kind: "platform/social-science" }
	| { readonly kind: "elective/own-course" }
	| { readonly kind: "elective/other-course" }
	| { readonly kind: "elective/other-faculty" }
	| { readonly kind: "unknown"; readonly raw: string };

export const SubjectCategory = {
	primary: (): SubjectCategory => ({ kind: "common-education/primary" }),
	liberalField: (field: FieldCategory): SubjectCategory => ({
		kind: "common-education/liberal/field",
		field,
	}),
	liberalForeignLanguage: (language: string): SubjectCategory => ({
		kind: "common-education/liberal/foreign-language",
		language,
	}),
	liberalCareer: (): SubjectCategory => ({
		kind: "common-education/liberal/career",
	}),
	// R6+ introductory course groups
	introCoreLearning: (): SubjectCategory => ({
		kind: "common-education/introductory/core-learning",
	}),
	introCoreEnglish: (): SubjectCategory => ({
		kind: "common-education/introductory/core-english",
	}),
	introForeignLanguage: (language: string): SubjectCategory => ({
		kind: "common-education/introductory/foreign-language",
		language,
	}),
	introMathAi: (): SubjectCategory => ({
		kind: "common-education/introductory/math-ai",
	}),
	// R6+ liberal course group: 7 fields
	liberalGroupLife: (): SubjectCategory => ({
		kind: "common-education/liberal-group/life",
	}),
	liberalGroupHealthSports: (): SubjectCategory => ({
		kind: "common-education/liberal-group/health-sports",
	}),
	liberalGroupCareer: (): SubjectCategory => ({
		kind: "common-education/liberal-group/career",
	}),
	liberalGroupArts: (): SubjectCategory => ({
		kind: "common-education/liberal-group/arts",
	}),
	liberalGroupHumanitiesSocial: (): SubjectCategory => ({
		kind: "common-education/liberal-group/humanities-social",
	}),
	liberalGroupNaturalScience: (): SubjectCategory => ({
		kind: "common-education/liberal-group/natural-science",
	}),
	liberalGroupComplex: (): SubjectCategory => ({
		kind: "common-education/liberal-group/complex",
	}),
	seminar12: (): SubjectCategory => ({ kind: "seminar/1-2" }),
	seminar34Spring: (): SubjectCategory => ({ kind: "seminar/3-4/spring" }),
	seminar34Fall: (): SubjectCategory => ({ kind: "seminar/3-4/fall" }),
	seminar56Thesis: (): SubjectCategory => ({ kind: "seminar/5-6-thesis" }),
	platformBasicA: (): SubjectCategory => ({ kind: "platform/basic-a" }),
	platformBasicB: (): SubjectCategory => ({ kind: "platform/basic-b" }),
	platformForeignLanguage: (): SubjectCategory => ({
		kind: "platform/foreign-language",
	}),
	platformAdvanced: (): SubjectCategory => ({ kind: "platform/advanced" }),
	// R6+ platform
	platformFacultyCommon: (): SubjectCategory => ({
		kind: "platform/faculty-common",
	}),
	platformHumanities: (): SubjectCategory => ({ kind: "platform/humanities" }),
	platformGlobalStudies: (): SubjectCategory => ({
		kind: "platform/global-studies",
	}),
	platformSocialScience: (): SubjectCategory => ({
		kind: "platform/social-science",
	}),
	electiveOwnCourse: (): SubjectCategory => ({ kind: "elective/own-course" }),
	electiveOtherCourse: (): SubjectCategory => ({
		kind: "elective/other-course",
	}),
	electiveOtherFaculty: (): SubjectCategory => ({
		kind: "elective/other-faculty",
	}),
	unknown: (raw: string): SubjectCategory => ({ kind: "unknown", raw }),
} as const;

export const isKind = <K extends SubjectCategoryKind>(
	category: SubjectCategory,
	kind: K,
): category is Extract<SubjectCategory, { kind: K }> => category.kind === kind;

export const isCommonEducation = (c: SubjectCategory): boolean =>
	c.kind.startsWith("common-education/");

export const isSeminar = (c: SubjectCategory): boolean =>
	c.kind.startsWith("seminar/");

export const isPlatform = (c: SubjectCategory): boolean =>
	c.kind.startsWith("platform/");

export const isElective = (c: SubjectCategory): boolean =>
	c.kind.startsWith("elective/");

export const isLiberal = (c: SubjectCategory): boolean =>
	c.kind.startsWith("common-education/liberal");

/**
 * Display names for categories: the single source of truth so diagnostics and
 * badges never surface a raw kind string (e.g. `common-education/liberal/field`).
 */
const KIND_DISPLAY_NAMES: Readonly<Record<SubjectCategoryKind, string>> = {
	"common-education/primary": "初年次科目",
	"common-education/liberal/field": "教養 分野",
	"common-education/liberal/foreign-language": "教養 外国語",
	"common-education/liberal/career": "教養 キャリア形成支援",
	"common-education/introductory/core-learning": "学びかた科目",
	"common-education/introductory/core-english": "基軸英語",
	"common-education/introductory/foreign-language": "初修外国語・日本語",
	"common-education/introductory/math-ai": "数理・データサイエンス・AI 科目",
	"common-education/liberal-group/life": "教養 生活",
	"common-education/liberal-group/health-sports": "教養 医療・健康・スポーツ",
	"common-education/liberal-group/career": "教養 キャリア形成",
	"common-education/liberal-group/arts": "教養 芸術",
	"common-education/liberal-group/humanities-social": "教養 人文・社会科学系",
	"common-education/liberal-group/natural-science": "教養 自然科学系",
	"common-education/liberal-group/complex": "教養 複合領域",
	"seminar/1-2": "ゼミナール I・II",
	"seminar/3-4/spring": "ゼミナール III（演習 I）",
	"seminar/3-4/fall": "ゼミナール IV（演習 II）",
	"seminar/5-6-thesis": "卒業論文・ゼミナール V・VI",
	"platform/basic-a": "PF 基礎 A 群",
	"platform/basic-b": "PF 基礎 B 群",
	"platform/foreign-language": "PF 外国語",
	"platform/advanced": "PF 発展",
	"platform/faculty-common": "PF 学部共通",
	"platform/humanities": "PF 人文科学分野",
	"platform/global-studies": "PF グローバル研究分野",
	"platform/social-science": "PF 社会科学分野",
	"elective/own-course": "自コース専門",
	"elective/other-course": "他コース専門",
	"elective/other-faculty": "他学部専門",
	unknown: "区分未判定",
};

export const kindDisplayName = (kind: SubjectCategoryKind): string =>
	KIND_DISPLAY_NAMES[kind];

File: crates/audit-domain/src/value/subject_category.rs
//! What bucket a course falls into. Ported from `value-objects/subject-category.ts`.
//!
//! [`SubjectCategory`] is the data-bearing category (some variants carry a field
//! or language). [`SubjectKind`] is its field-less discriminant — a `Copy` value
//! used for set membership, caps, and display, keeping "which bucket" separate
//! from "the bucket's payload". Only foreign-language variants carry a language
//! and only the liberal-field variant carries a field, so an illegal pairing
//! (say, a language on a seminar) is unrepresentable.

use serde::{Deserialize, Serialize};

use super::field_category::FieldCategory;
use super::language::Language;

/// The field-less discriminant of a [`SubjectCategory`]. Wire form matches the TS
/// `SubjectCategoryKind` union.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum SubjectKind {
    #[serde(rename = "common-education/primary")]
    CommonPrimary,
    #[serde(rename = "common-education/liberal/field")]
    LiberalField,
    #[serde(rename = "common-education/liberal/foreign-language")]
    LiberalForeignLanguage,
    #[serde(rename = "common-education/liberal/career")]
    LiberalCareer,
    #[serde(rename = "common-education/introductory/core-learning")]
    IntroCoreLearning,
    #[serde(rename = "common-education/introductory/core-english")]
    IntroCoreEnglish,
    #[serde(rename = "common-education/introductory/foreign-language")]
    IntroForeignLanguage,
    #[serde(rename = "common-education/introductory/math-ai")]
    IntroMathAi,
    #[serde(rename = "common-education/liberal-group/life")]
    LiberalGroupLife,
    #[serde(rename = "common-education/liberal-group/health-sports")]
    LiberalGroupHealthSports,
    #[serde(rename = "common-education/liberal-group/career")]
    LiberalGroupCareer,
    #[serde(rename = "common-education/liberal-group/arts")]
    LiberalGroupArts,
    #[serde(rename = "common-education/liberal-group/humanities-social")]
    LiberalGroupHumanitiesSocial,
    #[serde(rename = "common-education/liberal-group/natural-science")]
    LiberalGroupNaturalScience,
    #[serde(rename = "common-education/liberal-group/complex")]
    LiberalGroupComplex,
    #[serde(rename = "seminar/1-2")]
    Seminar12,
    #[serde(rename = "seminar/3-4/spring")]
    Seminar34Spring,
    #[serde(rename = "seminar/3-4/fall")]
    Seminar34Fall,
    #[serde(rename = "seminar/5-6-thesis")]
    Seminar56Thesis,
    #[serde(rename = "platform/basic-a")]
    PlatformBasicA,
    #[serde(rename = "platform/basic-b")]
    PlatformBasicB,
    #[serde(rename = "platform/foreign-language")]
    PlatformForeignLanguage,
    #[serde(rename = "platform/advanced")]
    PlatformAdvanced,
    #[serde(rename = "platform/faculty-common")]
    PlatformFacultyCommon,
    #[serde(rename = "platform/humanities")]
    PlatformHumanities,
    #[serde(rename = "platform/global-studies")]
    PlatformGlobalStudies,
    #[serde(rename = "platform/social-science")]
    PlatformSocialScience,
    #[serde(rename = "elective/own-course")]
    ElectiveOwnCourse,
    #[serde(rename = "elective/other-course")]
    ElectiveOtherCourse,
    #[serde(rename = "elective/other-faculty")]
    ElectiveOtherFaculty,
    #[serde(rename = "unknown")]
    Unknown,
}

impl SubjectKind {
    /// The canonical wire string (matches the TS kind union member).
    pub const fn as_str(self) -> &'static str {
        use SubjectKind::*;
        match self {
            CommonPrimary => "common-education/primary",
            LiberalField => "common-education/liberal/field",
            LiberalForeignLanguage => "common-education/liberal/foreign-language",
            LiberalCareer => "common-education/liberal/career",
            IntroCoreLearning => "common-education/introductory/core-learning",
            IntroCoreEnglish => "common-education/introductory/core-english",
            IntroForeignLanguage => "common-education/introductory/foreign-language",
            IntroMathAi => "common-education/introductory/math-ai",
            LiberalGroupLife => "common-education/liberal-group/life",
            LiberalGroupHealthSports => "common-education/liberal-group/health-sports",
            LiberalGroupCareer => "common-education/liberal-group/career",
            LiberalGroupArts => "common-education/liberal-group/arts",
            LiberalGroupHumanitiesSocial => "common-education/liberal-group/humanities-social",
            LiberalGroupNaturalScience => "common-education/liberal-group/natural-science",
            LiberalGroupComplex => "common-education/liberal-group/complex",
            Seminar12 => "seminar/1-2",
            Seminar34Spring => "seminar/3-4/spring",
            Seminar34Fall => "seminar/3-4/fall",
            Seminar56Thesis => "seminar/5-6-thesis",
            PlatformBasicA => "platform/basic-a",
            PlatformBasicB => "platform/basic-b",
            PlatformForeignLanguage => "platform/foreign-language",
            PlatformAdvanced => "platform/advanced",
            PlatformFacultyCommon => "platform/faculty-common",
            PlatformHumanities => "platform/humanities",
            PlatformGlobalStudies => "platform/global-studies",
            PlatformSocialScience => "platform/social-science",
            ElectiveOwnCourse => "elective/own-course",
            ElectiveOtherCourse => "elective/other-course",
            ElectiveOtherFaculty => "elective/other-faculty",
            Unknown => "unknown",
        }
    }

    /// Japanese display label (mirrors `KIND_DISPLAY_NAMES`).
    pub const fn display_name(self) -> &'static str {
        use SubjectKind::*;
        match self {
            CommonPrimary => "初年次科目",
            LiberalField => "教養 分野",
            LiberalForeignLanguage => "教養 外国語",
            LiberalCareer => "教養 キャリア形成支援",
            IntroCoreLearning => "学びかた科目",
            IntroCoreEnglish => "基軸英語",
            IntroForeignLanguage => "初修外国語・日本語",
            IntroMathAi => "数理・データサイエンス・AI 科目",
            LiberalGroupLife => "教養 生活",
            LiberalGroupHealthSports => "教養 医療・健康・スポーツ",
            LiberalGroupCareer => "教養 キャリア形成",
            LiberalGroupArts => "教養 芸術",
            LiberalGroupHumanitiesSocial => "教養 人文・社会科学系",
            LiberalGroupNaturalScience => "教養 自然科学系",
            LiberalGroupComplex => "教養 複合領域",
            Seminar12 => "ゼミナール I・II",
            Seminar34Spring => "ゼミナール III（演習 I）",
            Seminar34Fall => "ゼミナール IV（演習 II）",
            Seminar56Thesis => "卒業論文・ゼミナール V・VI",
            PlatformBasicA => "PF 基礎 A 群",
            PlatformBasicB => "PF 基礎 B 群",
            PlatformForeignLanguage => "PF 外国語",
            PlatformAdvanced => "PF 発展",
            PlatformFacultyCommon => "PF 学部共通",
            PlatformHumanities => "PF 人文科学分野",
            PlatformGlobalStudies => "PF グローバル研究分野",
            PlatformSocialScience => "PF 社会科学分野",
            ElectiveOwnCourse => "自コース専門",
            ElectiveOtherCourse => "他コース専門",
            ElectiveOtherFaculty => "他学部専門",
            Unknown => "区分未判定",
        }
    }

    /// Whether this is any common-education kind.
    pub fn is_common_education(self) -> bool {
        self.as_str().starts_with("common-education/")
    }

    /// Whether this is any seminar kind.
    pub fn is_seminar(self) -> bool {
        self.as_str().starts_with("seminar/")
    }

    /// Whether this is any platform kind.
    pub fn is_platform(self) -> bool {
        self.as_str().starts_with("platform/")
    }

    /// Whether this is any elective kind.
    pub fn is_elective(self) -> bool {
        self.as_str().starts_with("elective/")
    }
}

/// A course's category, carrying any payload the kind implies. Wire form is the
/// TS shape `{ "kind": "...", <field|language|raw>? }` (internally tagged on `kind`).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum SubjectCategory {
    #[serde(rename = "common-education/primary")]
    CommonPrimary,
    #[serde(rename = "common-education/liberal/field")]
    LiberalField { field: FieldCategory },
    #[serde(rename = "common-education/liberal/foreign-language")]
    LiberalForeignLanguage { language: Language },
    #[serde(rename = "common-education/liberal/career")]
    LiberalCareer,
    #[serde(rename = "common-education/introductory/core-learning")]
    IntroCoreLearning,
    #[serde(rename = "common-education/introductory/core-english")]
    IntroCoreEnglish,
    #[serde(rename = "common-education/introductory/foreign-language")]
    IntroForeignLanguage { language: Language },
    #[serde(rename = "common-education/introductory/math-ai")]
    IntroMathAi,
    #[serde(rename = "common-education/liberal-group/life")]
    LiberalGroupLife,
    #[serde(rename = "common-education/liberal-group/health-sports")]
    LiberalGroupHealthSports,
    #[serde(rename = "common-education/liberal-group/career")]
    LiberalGroupCareer,
    #[serde(rename = "common-education/liberal-group/arts")]
    LiberalGroupArts,
    #[serde(rename = "common-education/liberal-group/humanities-social")]
    LiberalGroupHumanitiesSocial,
    #[serde(rename = "common-education/liberal-group/natural-science")]
    LiberalGroupNaturalScience,
    #[serde(rename = "common-education/liberal-group/complex")]
    LiberalGroupComplex,
    #[serde(rename = "seminar/1-2")]
    Seminar12,
    #[serde(rename = "seminar/3-4/spring")]
    Seminar34Spring,
    #[serde(rename = "seminar/3-4/fall")]
    Seminar34Fall,
    #[serde(rename = "seminar/5-6-thesis")]
    Seminar56Thesis,
    #[serde(rename = "platform/basic-a")]
    PlatformBasicA,
    #[serde(rename = "platform/basic-b")]
    PlatformBasicB,
    #[serde(rename = "platform/foreign-language")]
    PlatformForeignLanguage,
    #[serde(rename = "platform/advanced")]
    PlatformAdvanced,
    #[serde(rename = "platform/faculty-common")]
    PlatformFacultyCommon,
    #[serde(rename = "platform/humanities")]
    PlatformHumanities,
    #[serde(rename = "platform/global-studies")]
    PlatformGlobalStudies,
    #[serde(rename = "platform/social-science")]
    PlatformSocialScience,
    #[serde(rename = "elective/own-course")]
    ElectiveOwnCourse,
    #[serde(rename = "elective/other-course")]
    ElectiveOtherCourse,
    #[serde(rename = "elective/other-faculty")]
    ElectiveOtherFaculty,
    #[serde(rename = "unknown")]
    Unknown { raw: String },
}

impl SubjectCategory {
    /// The field-less discriminant of this category.
    pub fn kind(&self) -> SubjectKind {
        use SubjectCategory as C;
        match self {
            C::CommonPrimary => SubjectKind::CommonPrimary,
            C::LiberalField { .. } => SubjectKind::LiberalField,
            C::LiberalForeignLanguage { .. } => SubjectKind::LiberalForeignLanguage,
            C::LiberalCareer => SubjectKind::LiberalCareer,
            C::IntroCoreLearning => SubjectKind::IntroCoreLearning,
            C::IntroCoreEnglish => SubjectKind::IntroCoreEnglish,
            C::IntroForeignLanguage { .. } => SubjectKind::IntroForeignLanguage,
            C::IntroMathAi => SubjectKind::IntroMathAi,
            C::LiberalGroupLife => SubjectKind::LiberalGroupLife,
            C::LiberalGroupHealthSports => SubjectKind::LiberalGroupHealthSports,
            C::LiberalGroupCareer => SubjectKind::LiberalGroupCareer,
            C::LiberalGroupArts => SubjectKind::LiberalGroupArts,
            C::LiberalGroupHumanitiesSocial => SubjectKind::LiberalGroupHumanitiesSocial,
            C::LiberalGroupNaturalScience => SubjectKind::LiberalGroupNaturalScience,
            C::LiberalGroupComplex => SubjectKind::LiberalGroupComplex,
            C::Seminar12 => SubjectKind::Seminar12,
            C::Seminar34Spring => SubjectKind::Seminar34Spring,
            C::Seminar34Fall => SubjectKind::Seminar34Fall,
            C::Seminar56Thesis => SubjectKind::Seminar56Thesis,
            C::PlatformBasicA => SubjectKind::PlatformBasicA,
            C::PlatformBasicB => SubjectKind::PlatformBasicB,
            C::PlatformForeignLanguage => SubjectKind::PlatformForeignLanguage,
            C::PlatformAdvanced => SubjectKind::PlatformAdvanced,
            C::PlatformFacultyCommon => SubjectKind::PlatformFacultyCommon,
            C::PlatformHumanities => SubjectKind::PlatformHumanities,
            C::PlatformGlobalStudies => SubjectKind::PlatformGlobalStudies,
            C::PlatformSocialScience => SubjectKind::PlatformSocialScience,
            C::ElectiveOwnCourse => SubjectKind::ElectiveOwnCourse,
            C::ElectiveOtherCourse => SubjectKind::ElectiveOtherCourse,
            C::ElectiveOtherFaculty => SubjectKind::ElectiveOtherFaculty,
            C::Unknown { .. } => SubjectKind::Unknown,
        }
    }

    /// The liberal-education field, if this is a liberal-field category.
    pub fn field(&self) -> Option<FieldCategory> {
        match self {
            SubjectCategory::LiberalField { field } => Some(*field),
            _ => None,
        }
    }

    /// The study language, if this category carries one.
    pub fn language(&self) -> Option<Language> {
        match self {
            SubjectCategory::LiberalForeignLanguage { language }
            | SubjectCategory::IntroForeignLanguage { language } => Some(*language),
            _ => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn kind_of_data_variants() {
        assert_eq!(
            SubjectCategory::LiberalField {
                field: FieldCategory::Humanities
            }
            .kind(),
            SubjectKind::LiberalField
        );
        assert_eq!(
            SubjectCategory::IntroForeignLanguage {
                language: Language::Chinese
            }
            .kind(),
            SubjectKind::IntroForeignLanguage
        );
        assert_eq!(
            SubjectCategory::Seminar56Thesis.kind(),
            SubjectKind::Seminar56Thesis
        );
    }

    #[test]
    fn payload_accessors() {
        let c = SubjectCategory::LiberalField {
            field: FieldCategory::Social,
        };
        assert_eq!(c.field(), Some(FieldCategory::Social));
        assert_eq!(c.language(), None);
        let lang = SubjectCategory::LiberalForeignLanguage {
            language: Language::German,
        };
        assert_eq!(lang.language(), Some(Language::German));
    }

    #[test]
    fn kind_groups() {
        assert!(SubjectKind::PlatformBasicA.is_platform());
        assert!(SubjectKind::Seminar12.is_seminar());
        assert!(SubjectKind::ElectiveOtherFaculty.is_elective());
        assert!(SubjectKind::CommonPrimary.is_common_education());
    }

    #[test]
    fn wire_shape_matches_ts() {
        let c = SubjectCategory::LiberalField {
            field: FieldCategory::Humanities,
        };
        assert_eq!(
            serde_json::to_string(&c).unwrap(),
            r#"{"kind":"common-education/liberal/field","field":"humanities"}"#
        );
        let u = SubjectCategory::Unknown {
            raw: "[謎]".to_owned(),
        };
        assert_eq!(
            serde_json::to_string(&u).unwrap(),
            r#"{"kind":"unknown","raw":"[謎]"}"#
        );
        assert_eq!(
            serde_json::to_string(&SubjectCategory::CommonPrimary).unwrap(),
            r#"{"kind":"common-education/primary"}"#
        );
    }

    #[test]
    fn kind_wire_string_matches() {
        assert_eq!(
            serde_json::to_string(&SubjectKind::Seminar56Thesis).unwrap(),
            "\"seminar/5-6-thesis\""
        );
    }
}
